### PR TITLE
Metrics stub

### PR DESCRIPTION
**Pull request by FrozenSake**
Metrics are an important piece of observability, and Prometheus is a standard default.
Exposes the /metrics endpoint, which will capture by default Prometheus go metrics.
Creates a new middleware for Metrics, and adds it to the default chain.

Note:
I am slightly unsure of my testing method for the attached middleware, as evidenced by the comment. There's a chance of it being flakey on a slow build server - part of the reason I use a .format() is because with milliseconds in the mix it was flakey about 50% of the time on my computer.  With the format, I have not been able to get it to flake out.

### CODE REVIEW

**Comment by FrozenSake** [comment]
Ah - I just noticed I missed following the proper format for commits. Should have been `go middleware: ...` and `go graphqlserver: ...`, and I guess `go mod: tidy`

If we'd like I can redo the PR fresh on a new branch to get good names, since doing it early for consistency is important, and reverting commits to change names is a pain.

**Comment by zdylag** [comment]
I think you can fix the commits but doing a "rebase -i". You can just rename them there?

File: middleware/metrics.go
Comment (zdylag) [inline middleware/metrics.go:37]: Why is this here and not after the "next" call?

File: middleware/metrics.go
Comment (zdylag) [inline middleware/metrics.go:45]: Is it useful to have this on the context?

File: middleware/metrics_test.go
Comment (zdylag) [inline middleware/metrics_test.go:15]: I bet we can time this a bit better by having one test where the internal handler takes no time, and assert it takes less than 5 seconds.

Then we can have another internal handler that takes 6 seconds and assert it takes more than 5 seconds.

Does that sounds easier to deflake?

File: middleware/metrics.go
Comment (FrozenSake) [reply middleware/metrics.go:37]: You raise a good point. I was thinking that having a way to get the current duration of a running request would be useful for observability, but it's not really clear that it is. I think that requires a rethink, but moving the .Observe call to right after the next.Serve will handle what I was trying to do in a non-janky way.

My mind went off on a tangent about making a HistogramVec (collection of Histograms) that we can access via this call. Something to think about.

File: middleware/metrics.go
Comment (FrozenSake) [reply middleware/metrics.go:45]: I think it depends on if we end up passing downstream, and pass a context down and back up. Something like a request that ends up calling a database, and passing the request metrics middleware all the way down - but reviewing that assumption, it's not clear that gives us more value than having downstream components basically time themselves (timestamping before, observing after) and report just that.

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:15]: Yeah, I think even less than 2 seconds shouldn't be flakey, given that having it at "equal" with second granularity I couldn't get it to flake any more. Without the format call, it was about 50/50 flakey or not.

**Comment by FrozenSake** [comment]
Honestly, looking at this, I think I need to rewrite it as a middleware that's attached to the mux, Lemme do that.

**Comment by zdylag** [comment]
@FrozenSake why does this need to be on the mux? Sounds okay as well, just curious what benefits you're seeing.

**Comment by FrozenSake** [comment]
Allows us to collect aggregate metrics a bit easier, simplifies managing collections, and keeps the logic a little easier to maintain. It also seems to be a more standard way that other people are handling it from poking around at solutions. Largely though, I think having it on the Mux simplifies the implementation details downstream. Could be wrong and willing (and happy) to be so. :)

File: app/graphql_server/main.go
Comment (zdylag) [inline app/graphql_server/main.go:53]: Can you move the rest of the middleware here?

Can you just add all of the middleware? If so do we even need the chain middleware?

File: middleware/metrics.go
Comment (zdylag) [inline middleware/metrics.go:61]: Do we need to wrap this? Can we not read rw.Header() after write?

File: middleware/metrics_test.go
Comment (zdylag) [inline middleware/metrics_test.go:32]: This is a super helpful test! Can we have a test for the middleware as well?

File: middleware/metrics.go
Comment (zdylag) [inline middleware/metrics.go:26]: I haven't used Prometheus in go, but is it standard to make so many top level vars?

**Review by zdylag — COMMENTED** [review]
Thanks for adding gorilla! Looks great!

File: middleware/metrics.go
Comment (zdylag) [reply middleware/metrics.go:61]: I just checked this and it seems like this is one reason I see people using the `negroni` package, to solve this exact issue haha. Can we use that instead?

File: middleware/metrics.go
Comment (zdylag) [reply middleware/metrics.go:26]: Looking around it seems like it might be. If so I wonder if it would be cleaner to put the metric declarations in their own file for each package? The registration can also be done in the init() I think. 

Open to other thoughts here. Just imagining what this looks like when we have many places with metrics.

Comment (FrozenSake) [reply middleware/metrics.go:26]: I was gonna say "as far as I can tell, yes," haha. I had taken them out of the init because I wasn't sure we wanted the registers to be called just from importing the middleware. No real preference either way.

Cleanliness wise, it's hard to say. Having the metrics definitions in their own file makes them less obtrusive, but also adds a layer of abstraction to understanding the code. Same as above, no real preference. I guess if pushed, I generally default towards things that are used together living together, but that might get real messy.

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:32]: Yeah, for sure! Let me try to whip something up later today.

File: middleware/metrics.go
Comment (FrozenSake) [reply middleware/metrics.go:61]: For sure, I'm not super familiar with the negroni package (it looks really good though), do you have an example of what you're thinking here I can reference?

Comment (zdylag) [reply middleware/metrics.go:61]: https://stackoverflow.com/a/63896952
^This is an example of exactly the change I think haha.

File: app/graphql_server/main.go
Comment (zdylag) [reply app/graphql_server/main.go:53]: It seems we can add middleware via Negroni instead?

https://github.com/urfave/negroni#with

^This seems like the more standard way to middleware that I am seeing (with gorilla used for routing mainly). If you don't mind can we move to this for middleware registration?

Comment (zdylag) [reply app/graphql_server/main.go:53]: (If not I can open a new PR after this one to do that move, no worries!)

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:40]: Okay, got one running. :)

File: app/graphql_server/main.go
Comment (FrozenSake) [reply app/graphql_server/main.go:53]: All about it, let me look at these patterns and implement. 

Part of why I'm getting in early is because my company has robust stuff in place, and I wanna fiddle with it from the start 😆 (and build my go dev experience that way)

Comment (FrozenSake) [reply app/graphql_server/main.go:53]: Okay, using Negroni will involve a bit of a rewrite of the handler funcs to match the `negroni.Handler` interface:
```go
type Handler interface {
  ServeHTTP(w http.ResponseWriter, r *http.Request, next http.HandlerFunc)
}```
I'm able to do that on my handlers, but I think doing it on the existing ones in my PR breaks the spirit of atomic PRs and makes roll back really hard. Should I make a second PR to change the existing right first, or do it all in one, do you think?

Comment (zdylag) [reply app/graphql_server/main.go:53]: I think doing the rest of the move on a follow up PR sounds fine yeah 👍

Comment (FrozenSake) [reply app/graphql_server/main.go:53]: Sweet. Now just to finish trying to wrangle Gorilla, Negroni, and Prometheus together to play nicely. 🤣

File: middleware/metrics.go
Comment (FrozenSake) [reply middleware/metrics.go]: I think I did what was being looked for here. The route/path extraction with mux is used because we need to get them for prometheus labels, and negroni call replaces the call I had here, and removes 3 blocks of boilerplate I had to write to enable it, so that's a good save.

File: app/graphql_server/main.go
Comment (FrozenSake) [reply app/graphql_server/main.go]: Okay, I couldn't get negroni to work properly. We poked at it in #backend-dev a bit as well, but ultimately something about the patterns here didn't "click" for me.

**Review by RageCage64 — APPROVED** [review]
This is kind of a soft approval, since I want @zdylag to have another chance to look before merging.

File: middleware/metrics_test.go
Comment (zdylag) [inline middleware/metrics_test.go:85]: nit: Let's not directly panic wherever possible.

File: middleware/metrics_test.go
Comment (zdylag) [inline middleware/metrics_test.go:83]: Does this ever exit? I feel like we would be leaking these http servers.

**Review by zdylag — APPROVED** [review]
Approving since this makes forward progress on the repo, but I think we will want to change that test to not spawn http servers.

File: middleware/metrics_test.go
Comment (BraydonKains) [reply middleware/metrics_test.go:85]: Is a `log.Fatal` equivalent prefferrable to a panic?

Comment (FrozenSake) [reply middleware/metrics_test.go:85]: Panic is generally just bad in general in go, and it's a bad on me to have included one.  Generally [effective go recommends against panic.](https://golang.org/doc/effective_go#panic)

So I consider this not a nit, and will look in to fixing it. I think a t.FailNow() is much better here.

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:83]: The go routine closes when the outside function ends.

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:47]: Still not against Init(), basically just need to decide if we're okay with calling prometheus registrations when we import the package, and if so, I'll move them there. It actually massively simplifies everything. Although since these are supposed to be specially constructed for this test, would still need an init here. :P

File: go.mod
Comment (FrozenSake) [reply go.mod:10]: That would mean this PR is the right place, so if we roll this back, it takes out the go mod changes (since this PR introduces the three imports)? If that's what you mean - I'm all for it.

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:83]: This may not be as clean as I thought, actually. It definitely cleans itself up once `go test` exits, but it's not clear the server closes before then, which would make other tests... not good.

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go]: t.failnow() won't work inside a goroutine, so I moved to an assert,fail with a go routine exit via `runtime.Goexit()`

File: middleware/metrics_test.go
Comment (zdylag) [reply middleware/metrics_test.go:84]: Naked goroutines do not exit when they leave scope. Only when they stop running (or the process dies). An alternative here would be to use `httptest.NewRecorder()` like we do in `graphql_server\server_test.go`. Then you don't need this goroutine at all.

Comment (FrozenSake) [reply middleware/metrics_test.go:84]: Okay, tested it with this code block:
```go
package main

import (
	"fmt"
	"time"
)

func main() {
	go func() {
		count := 0
		for {
			fmt.Printf("Hello, playground, %d\n", count)
			count++
		}
	}()
	time.Sleep(5 * time.Nanosecond)
	fmt.Println("Going to dead time!")
}
```
It exits cleanly and doesn't keep running in the background.

Comment (FrozenSake) [reply middleware/metrics_test.go]: I got learned, and switched it to a httptest.Server, and use a defer close so we don't leave it running.

**Comment by FrozenSake** [comment]
Just to check, anything else we want to address in the moment? Looks like outstanding decisions are:
`Init() versus calling a special Prometheus Init function` --> Restating: I'm indifferent, only reason for the separate function was to not always register Prometheus when importing middleware. `Init()` is much simpler, but less control.
`Rebasing go mod commits into the commits with the imports`
`should there be a metrics definition file because we declare a lot of top level vars in golang prometheus`

File: middleware/metrics_test.go
Comment (zdylag) [inline middleware/metrics_test.go:81]: No more go! 🎉

**Review by zdylag — APPROVED** [review]
All of the remaining comments are very much things we don't need to deal with in this PR. Let's merge it and unblock people!

File: middleware/metrics_test.go
Comment (FrozenSake) [reply middleware/metrics_test.go:81]: It's go(ing), go(ing), go(ne).

**Pull request merged**
